# Fiass_Database or DB : Facebook AI Similarity Search

# Install environment : create an enviornment or env

- it will craete directory. or it simply create the folder
- Below step I can follow : Below code copy and paste inside terminal and hit enter.
- here -> (-n) means name of folder, I given name as "vectordb"

- code :  [ conda create -n vectordb python=3.10 ] 
- Now, activate the env -> code : [ conda activate vectordb ] 
- Here means -> conda activate & give the folder name or env name -> "vectordb"

# Install libaries :

- Dont run here, it will not work my case, open the terminal & run inside it, its working copy and paste these below libaries.

- 1.pip install ipykernel
- 2.python -m ipykernel --version  <- this is basically to check the version
- 3.import faiss <- import the Faiss library, which stands for Facebook AI Similarity Search.
- 4.pip install faiss-cpu 
- 5.pip install numpy
- 6.pip install requests

# Note : Just for understanding !
- Select the Kernel inside VS code
- 1.If I will pick -> venv/python.exe <- this is my present directory if I will select this then, it will not the "ipykernel".

- 2.If I will pick these both any one option -> "conda\envs\venv\python.exe" | "base\conda" then both option works, because its Global.
- So, I will pick the -> venv( python 3.18.10 )\"conda\envs\venv\python.exe"

In [22]:
pip install numpy

^C
Note: you may need to restart the kernel to use updated packages.


In [ ]:
pip install ipykernel

Note: you may need to restart the kernel to use updated packages.


In [1]:
import faiss

In [ ]:
pip install faiss-cpu

   ---------------------------------------- 0.0/18.2 MB ? eta -:--:--
   - -------------------------------------- 0.5/18.2 MB 3.4 MB/s eta 0:00:06
   -- ------------------------------------- 1.3/18.2 MB 3.4 MB/s eta 0:00:06
   ----- ---------------------------------- 2.6/18.2 MB 4.6 MB/s eta 0:00:04
   --------- ------------------------------ 4.2/18.2 MB 5.6 MB/s eta 0:00:03
   ------------- -------------------------- 6.3/18.2 MB 6.3 MB/s eta 0:00:02
   ----------------- ---------------------- 7.9/18.2 MB 6.6 MB/s eta 0:00:02
   -------------------- ------------------- 9.2/18.2 MB 6.6 MB/s eta 0:00:02
   ----------------------- ---------------- 10.5/18.2 MB 6.5 MB/s eta 0:00:02
   ------------------------- -------------- 11.5/18.2 MB 6.3 MB/s eta 0:00:02
   --------------------------- ------------ 12.6/18.2 MB 6.2 MB/s eta 0:00:01
   ------------------------------ --------- 13.9/18.2 MB 6.1 MB/s eta 0:00:01
   -------------------------------- ------- 14.9/18.2 MB 6.0 MB/s eta 0:00:01
 

In [ ]:
pip install requests

  Using cached requests-2.32.5-py3-none-any.whl.metadata (4.9 kB)
  Using cached idna-3.10-py3-none-any.whl.metadata (10 kB)
  Using cached urllib3-2.5.0-py3-none-any.whl.metadata (6.5 kB)
  Using cached certifi-2025.10.5-py3-none-any.whl.metadata (2.5 kB)
Using cached requests-2.32.5-py3-none-any.whl (64 kB)
Using cached idna-3.10-py3-none-any.whl (70 kB)
Using cached urllib3-2.5.0-py3-none-any.whl (129 kB)
Using cached certifi-2025.10.5-py3-none-any.whl (163 kB)

   ---------------------------------------- 0/5 [urllib3]
   ---------------------------------------- 0/5 [urllib3]
   ---------------------------------------- 0/5 [urllib3]
   ---------------------------------------- 0/5 [urllib3]
   ---------------------------------------- 0/5 [urllib3]
   ---------------------------------------- 0/5 [urllib3]
   -------- ------------------------------- 1/5 [idna]
   -------- ------------------------------- 1/5 [idna]
   ---------------- ----------------------- 2/5 [charset_normalizer]
   

# Note :

- Now everything is installed, now we need a data, we are going to perform some sort of Embedding Operation. 
- "Embedding Operation" - means we are going to take some textual information, and eventually we will try to convert those textual 
- information, into a numerical vector. After doing that so, we are going to store those information into a databases.
- This is something we will going do over here. 
- We need a dataset, I can get the data from the multiple sources. 
- Again in realtime, people will going to give you the data again from multiple different different places, there is possibility that,

- people will going to provide the data into a
- 1.PDF format,
- 2.Excel sheet,
- 3.URL - & asked you to scrapped those data, may be data will be avaiable in some SQL or NoSQL databases.

- There is alot of possibiity that, dataset will be avaiable in different diffent location.

# Next :

- As of now we will discussed the Conversion the data and storage part, but in general in realtime,
- we try to build a complete Datapipeline. So, where we try to read data from the multiple sources. There is possibility that I am going 
- to received the data into a batch mode. There is possibility that I will be received a data into a "streaming mode" & there is also a 
- possibility that I might received a data into a "mini batch mode", means I am talking about the frequency of the data.
- There is the possibility & there is possibility that data is coming from multiple different different sources. Maybe from the 

- 1.S3 Bucket, 
- 2.Azure Blobs 
- or maybe from the databases, there are 100 of different different databases, whihc is avaiable maybe from the different different
- file system.

- As of now I am going to take the data from this page -> Steps : 
- Go to "Euron", next go to "Founder story" URL : https://euron.one/our-founder

# Next :

- Lets come this page and copy both paragram from this : "Making an Impact".  URL : https://euron.one/our-founder

- 1.Helping Millions of Students Succeed & <- Copy the paragraph 
- 2.The Entrepreneur and Teacher: Sudhanshu's Dual Legacy <- Copy the paragraph 

- So, lets create the "vaiable = data" &  Paste the data, that I have copied from the these 2 paragraph.
- so, this is the data that,I am copy and paste from the website.
- Also there is possibility that we will received a data from the PDF format. In that case we have to write the Python Function to read
- out those data from a PDF file. Maybe someone going to give the data into a 

- 1.Word format
- 2.Excel sheet format
- 3.( .Txt format ) 
- 4.JSON format

- So, there can be thousand of different different kind of sources of the data, that we will be able to find out, in a realtime.

# Next : 

- If we see this below data, then we can observed that, there is different different lines or paragraphs. 
- So, my very 1st approach should be to break down, this entire dataset into a small, small chunks. 
- For example : We are reading a PDF, which is 100 of pages somewhere around, so its not like I will take one single variable and I will
- try to keep all those hundreds of pages, and then I will try to convert those 100 of pages, all together in one single vector.
- Because that is not going to work for me, Because the whole idea behind doing "Embedding" is to improvise our Search. 
- Improvise our Indexing, that is not going to help.
- So, what I am suppose to do next, If I have this kind of dataset,
- then basically I will try to break this data, into a small, small chunks of size may be 500 text or maybe 1,000 of texts. Or maybe 
- thousand characters, or maybe thousand of words, depends.
- But, Yes I am going to break down this into a small, small chunks, by keeping some sort of a overlap between a chunks and
- why we are going to maintain that.

# Let's talk about this

- So we are going to take this below data.

In [43]:
data = """ 
Making an Impact
Helping Millions of Students Succeed
Sudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.
Many of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.
In 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron continued to offer some of the most affordable and accessible tech courses in the world.

The Entrepreneur and Teacher: Sudhanshu's Dual Legacy
Sudhanshu's journey isn't just one of entrepreneurial success; it's also a story of dedication to teaching. Throughout his career, he has remained a passionate educator, constantly looking for ways to empower others through knowledge. Whether teaching courses in Big Data,
Data Science, or programming, Sudhanshu has always sought to make complex subjects accessible to learners at all levels.

His commitment to affordable education has earned him the respect and admiration of countless students. Many credit Sudhanshu with changing their lives, helping them secure jobs, improve their skills, and break free from the limitations of their backgrounds."""

# Next :

- Now here we are going to call a -> "data.strip()" <- This will eliminate the all those leading and trailing spaces inbetween our data. 
- Let me name it as a "Clean_data" a Variable name.
- There so many different types of clearning we can try to perform, this is just a symbolic one.
- We can perfrom a Normilization operation, we can just try to remove the pencutations, or any kind of unncessary most of the time 
- a repetitive words. So, many differen different type kind of NLP based cleaning we can try to perform on top of data, now but again that
- depends upon the kind of problem statement, which I am going to solve. So that depends more like on a business domain, or like problem
- statement that we are going to solve.
- So, as of now just "Symbolic" manner so I have called "data strip", so which going to remove leading & trailling spaces from the dataset.
- I am calling that dataset as a clean data lets suppose.
- But again, on top of this we do a lot of processing with respect to our data. 

# Next : 

- Let's suppose, I a trying to read a PDF, there will be so many different different kind of edge case, getting & I have to handle, all
- of the edge cases, then only I will be able to get the required dataset, but here dataset that we are using is not much complicated,
- There is a possibility, that I will be having some of the word or some of the line inside my dataset into a different language,
- May be Into a "Spanish", "Franch" or "Hindi" language, there will be very high chance, when I am trying to read some sort of a books,
- or maybe a PDF, or some sort of a data, there is chance that I am end of getting, mixture of a language, which we say multi lanuage.
- there is a very high possibility, I have to handle all of those situation, with respect to a dataset, then only I can try to make my  
- dataset ideal or any kind of operation, whether its a embedding or maybe whether a "RAG" or any other agentic operation. 
- alots of thing is goes around interms of building a complete data pipeline, again that is major roles and responsibility we all try to
- inside a organization.

- So, here I am able to read a data in a Symbolic manner, I am able to call a "data.Strip()". 

In [3]:
clean_data = data.strip()

In [4]:
clean_data

"Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.\nMany of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron continued to offer some of the most affordable and accessible tech courses in the world.\n\nThe Entrepreneur and Teacher: Sudhanshu's Dual Legacy\nSudhanshu's journey isn't just one of entrepreneurial success; it's also a story of dedication to teaching. Throughout his career, he has remained a

# What is the total length of my string or paragraph?

- As a outcome, we can see its a -> ( 1,497 ).
- This is basically total, like a space which my dataset has occupied.

In [5]:
len(clean_data)

1497

# Next : 

- Now my requirement is breakdown this dataset, into a small small chunks, means 
- For example :
- 1.One single line, I would like to keep maybe 800 character, that is one chunks.
- 2.Another 800 character this is again another chunk.
- 3.Another 800 character this is again another chunk.

- Like this I can try to create a chunks, with certain chunks size, so in one single sentences, or one single data, having only that much
- number of characters or may be I can try to define it, based on the words, whatever suits us. 

# Next: 

- Let me define some of the variables :

- Max_char or maximum character = 800, means in one single line there will be a 800 character max to max, not more then that.
- Then I can try to define "Overlapp" = 100, means lets suppose I have line No.1 & I have line No.2, there is a overlapped of 100 character
- between the line No.1 and line No.2. 

- Code : [ overlap = 100 ]
- Why I am trying to keep the overlapp here ?
- Because when I am going to do a chunking operation, when I am going to breakdown this data or paragraph into a chunks.
- So, I have to maintain even a relation, means context. If there will be an overlapped, between a data, onviously it will be able to see
- the contect, it will be able to correlate.  And when I am going to do search operation, on the top of the embedding, it will be able to
- give me like, there is some match between line No.1 and there is some match between Line No.2 as well. Just to maintain that context,
- I am going to maintain the overlapping, between 2 datasets. That is the region, we have given overlapped, with hundreds of like
- a characters.

- Code : [ Chunk = []  ] <- Then try to create a chunks as a "List"
- Then I will try to keep [i] variable as a [0] zero. means start from [0] indexes. Here -> "Making an Impact" in paragraph we can see.

- Code : [ While i <len(clean_data): ]
- while [i] is lesser (<) than, length of a (clean_data), start from [0] index go till last one, this is the region lenth of (Clean_data)
- If I have 500 word or more then that in my pagagraph it will go till last, then only it will going to break the loop,this is what I mean
- While i <len(clean_data): <- when this statement is true, then it has to perform different different operation, 

- What kind of operation, it has to perform? 
- it has to break down my entire dataset into a small, small pieces.

- Code : piece = clean_data[i:i+max_char] 
- Lets suppose, here I am going to call my "Clean_data" , "Clean_data" is my variable name by the way, my whole paragraph is inside
- this variable.
- "Clean_data"[] <- inside this I am going to write the index, like start from here and then go till this particular point.
- So, I am performing slicing opeartion, I am saying start from [i] means start from [0] index, go till [i] + maximum character which is
- our 800. means [0+800]. this -> clean_data[i:i+max_char] whole will going to be one data.
- Now, lets say I will keep this data inside varaiable -> ("Piece"). This -> clean_data[i:i+max_char] <- will be having in a very 
- 1st interation, it will be having a data (i = 0th) index till 800, this is the maximum character which I have given.
- This is the line No.1 which I am going to maintain,  
- 
- Code : chunks.append(piece)
- Here, I am going to append "chunks", what ? append ("piece"). this is the list -> [ Chunks = [] ] <- which I have created.
- So, I am going to append this -> Code : piece = clean_data[i:i+max_char]  & chunks.append(piece) <- particular piece inside this
- particulat list -> [ chunks = [] ] <-

- Code : i = i+max_char - overlap 

- Now we will going to increment the value of [i], we have to shift the value of [i] now.

- Now [i] is = [i]+max_char-overlap  
- Now, I am going to reset my value of [i], where will be my [i] by the way?
- so, I has started from the [0] zero, it went till the 800, now my [i] is avaiable in the 800th location.
- But, I have to set the value of this [i] at the (700) location.why?
- So that, I will be able to maintain a overlap, of 100 inbetween. Between from the 1st sentences & between the 2nd sentences. 
- That is the region, I have written the (i+max_char) means ( 0+800 ) - 100. which means the 700, overlap.
- SO, now my [i] is here -> in 700.  [ 0,1,2,3,4,5,6.........700,701,702,703......800 ]
- so, next time, when [i] is going to start from where? this line code[i], I am talking about here -> while i<len(clean_data):
- Next time, [i] will going to start from the 700 by the way, [700, 7001, 7002......800 ], then (700+800) = 1500 technically.
- so, it will go till the (1500). 
- Again, I will try to reset the value of [i] at (1400). then again I will going to start from (1400), then I will be able to maintain
- overlapped of 100, 100, 1000 inbetween. 
- Here, I am playing the value of [i], with those [i] those indexes, I am able to maintain the overlapped.


In [6]:
max_char = 800
overlap = 100
chunks = []
i = 0
while i <len(clean_data):
    piece = clean_data[i:i+max_char]
    chunks.append(piece)
    i = i+max_char - overlap

# Why do we need to maintain the overlapped?
- we need to maintain the overlapped, so that we will be maintain the context, a little bit of context we will be able maintain in the 
- both the sentences.
- The total is 1497, like a space which my dataset has occupied. so max to max I am able to creating 2 chunks. 
- Inside the chunks, I will be having the max to max like 2 sentences or 3 sentences.

In [7]:
chunks

["Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.\nMany of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron continued to offer some of the most affordable and accessible tech courses in the world.\n\nThe Entrep",
 "continued to offer some of the most affordable and accessible tech courses in the world.\n\nThe Entrepreneur and Teacher: Sudhanshu's Dual Legacy\nSudhanshu's journey isn't just one of en

# check Lenth or (len)

- Here, If I will going to check his length. 
- As we can see its a 3 basically. So, max to max 3 sentences I am able to form, not more then that.  

In [8]:
len(chunks)

3

# Next : [ max_char = 300 ]

- Lets say, I am taking or maintain as [ max_char = 300 ].
- As a outcome, we can see over here, I am able to create 8 lines, out of my dataset & all of these lines are avaiable inside this
- particular list[].

- If we observed here, in our outcome, we can see some symbols, (\), like this we have to clean in the cleaing operation, because these
- will not going to give me value, with respect to a data. We have to perform some claning opeartion on top of this.
- There maybe some numerical value which is not of our use.
- Maybe our PDF data is avaiable in some different language, so then again we have to go with the cleaning operation. 
- There are alot of "Data preprocessing" operation, we try to perform, which is true with across all the application.
- Not just with this application, whether we are trying to build a simple "Machine Learning" model or "Deep Learning" model.
- or we are dealing with some sort of "Data analytics" everywhere, we have to go through this kind of process.

# Next, : Overlapped

- In this dataset, there are overlapped also, we we cross check here, if we go end of this data 
- -> this 1st line inside outcome, we can see -> "skills they need to succeed in toda ". ("toda") <- this only it is able to capture 
- Here, we will be able to see that 2nd paragraph of this outcome -> ["skills they need to succeed in today's competitive job market"] 
- Here, we can see full name -> Today's not "toda".
- There is overlapped, I am trying to maintain between 2 sentences, so that it will be able to maintain a contextual information.
- It will be able to understand something, like this 2 lines are connected to each others.

# Next :

- Now, we are able to understand how we are able to create an chunks, simple Python function.
- There are inbuild libaries also there, which we can even try to use that one. Moreless this is something people are doing currently.
- even insid those libaries.

In [9]:
max_char = 300
overlap = 100
chunks = []
i = 0
while i <len(clean_data):
    piece = clean_data[i:i+max_char]
    chunks.append(piece)
    i = i+max_char - overlap

In [10]:
chunks

["Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in toda",
 "1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.\nMany of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 202",
 'backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron ',
 "gnificant milestone, Sudhanshu remained focused on his mission. Even after the

In [11]:
len(chunks)

8

# Note : Convert this entire dataset into a Embedding.

- Convert this entire dataset into a Embedding, into a numerical format which we have to convert this data.
- We done last class, we can use "EuriAPI", we can use some models from the "Hugging Face" or any open sources model, which is going 
- to support a embeddding and I will be able to convert my dataset into a vector. 

- Let's try to convert this entire dataset into a embedding, into a vector spaces.

# Note : Follow the steps : Just copy and paste dn't run below code.

- Go to the "EuriAPI" -> Code Example -> Text Embedding Similarity <- Just copy the code from here. This is basically for "Text Embedding Similarity"

- Just copy and paste the code, dont run it, we need to replace with "API keys" and "Authorization". 

- Without "API keys" we cant use this piece of code.

In [ ]:
#import requests
#import numpy as np

#def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer ********************"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

#text = "The weather is sunny today."

#embedding = generate_embeddings(text)

# Lets start :

- To get API key : Follow steps : Go to "EURI" and -> ["API keys"] -> Name it & create it. I named it as -> ["Fiass_Vector_DB API"]
- This is my API Key :  [   euri-78c66f791ec6c9b285bc2f92d2487d5c61f0bec3c10c01a8f89fb20aba1c5ae1   ]

- Just replace this API key here -> [ "Authorization": "Bearer ********************" ]-< 

# Next :

- This is the function, -> [def generate_embeddings(text): ] <- which will going to take the data, a text and it is going 
- give the Embedding. It is going to take the data and it is going to give the Embedding.
- we replace the API, we will clean and remove last 2 lines code & just execute the below code.

In [12]:
import requests
import numpy as np

def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer euri-78c66f791ec6c9b285bc2f92d2487d5c61f0bec3c10c01a8f89fb20aba1c5ae1"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

# Next :

- Now we have the data, we have the "Chunks", from this Chunks, I have to convert this entire data into a embedding by using, this 
- "EURI API" 
- Now here this one -> [ "model": "text-embedding-3-small" ] <- we are using the OpenAI based model, again this is the multilingual model.
- that we are using, it support hundreds of languages, bydefault. When people was training this model obviously they have used the 
- multilingual dataset. 
- But, again one of the lighted & fasted embedding model and one of the low cost embedding model, we will be able to find out
- in the market. There are again hundreds of other embedding model which is avaiable. 

- For Example :
- Last class we have used from Hugging Face a "Qwen3-Embedding-8B" model. So, we can even go with that.

- So, this model -> [ "model": "text-embedding-3-small" ] will going to give us a vector size return of ( 1536 ) is the dimension of the
- one single vector, it is going to return.


# Next :

- So, we have a "Chunks". we have a data, we will just run my generate embedding through this chunks, one by one & eventually,
- I will be able to get my final embedding. 

- So, here , I am going to write the code, 

- code : [ "for i in chunks": ]
- code : [ embedding = generate_embeddings(i) ] 
- code : [ print(embedding) ]
- one by one lets do the Chunking & call this generate embedding function, -> [ embedding = generate_embeddings(i) ] <- pass the value 
- one by one, then try to print the emdedding. If will try to keep all embedding somewhere. 
- Now lets execute below code, As a outcome we can see that, it is like creating embedding for sentences,
- for sentences No.1, Sentences No.2,Sentences No.3, Sentences No.4, Sentences No.5, Sentences No.6 etc...... 
- So, I have total 8 sentences. For all of those sentences it will able to create the embedding. So, it is able to create the numerics.

In [13]:
for i in chunks:
    embedding = generate_embeddings(i)
    print(embedding)

[ 0.01636502 -0.01330146  0.04360536 ... -0.00716132 -0.00025957
  0.00777273]
[-0.02828987 -0.00211549  0.03156904 ... -0.00953185 -0.01975844
 -0.00468256]
[-0.01842189 -0.03448532  0.01168446 ... -0.0130075   0.00758591
  0.02325387]
[-0.00206819 -0.01932828  0.04296086 ... -0.033941    0.00859491
  0.02564766]
[ 0.01036524 -0.00490744  0.05058131 ... -0.0313709  -0.00884518
  0.00210155]
[-1.0751131e-02 -2.5514962e-02  3.7223070e-02 ... -1.7212370e-02
 -6.2131260e-05  4.0052750e-04]
[ 0.00257804 -0.02860213  0.04973948 ...  0.00047777 -0.02138105
  0.00431918]
[ 0.04722496  0.00496422  0.04577735 ... -0.01690703 -0.07669617
  0.00453403]


# Next :

- Once I will be able to generate this chunks, now we can happily store this data into any of these databases.

- [ 0.01636502 -0.01330146  0.04360536 ... -0.00716132 -0.00025957
-  0.00777273]
-[-0.02828987 -0.00211549  0.03156904 ... -0.00953185 -0.01975844
- -0.00468256]
-[-0.01842189 -0.03448532  0.01168446 ... -0.0130075   0.00758591
-  0.02325387]  etc.......

- So, whether it's an Quadrant, whether its an "Pinecone" wherther it's "Chroma DB", whether it's a "Facebook AI Similarity Search".
- or any other databases, which is exist in my local or maybe which is available in some other platforms, depends, I will be able to 
- store this dataset in anywhere that I want. 

# Note : "Enumerate" ?
- Lets understand what does "Enumerate" basically means inside python?

- If we trying to pass a "List", so Enumeration will try to give a ID. 
- Lets suppose, if we have a List[] inside this list we have sentences -> ["vijay", "Euron", "iNeuron" ] <- this kind of data, we have
- Let's try to write "for i in "Enumeration" & print(i).
- so, as a outcome, we can see over here, what it does basically. whatever data that we have inside the List["vijay", "Euron", "iNeuron"],
- it will take this data and it will try to return, indexes as well.(0,1,2).
- This is what "Enumerate" does.


In [14]:
l = ["vijay", "Euron", "iNeuron" ]

In [15]:
for i in enumerate(l):
    print(i)

(0, 'vijay')
(1, 'Euron')
(2, 'iNeuron')


# Next :

- So here we just generate the embeddings, & I am just trying to like show the print, in a ideal way.
- So, I have to basically create a complete embedding list and I have to create a Meta information, & then I have to go ahead and store,
- those data. So, lets remodify these above piece of code. 

- Lets write code this way :
- For idx & chunks in enumerate (chunks):
- This is just extension of above code that we are writting. Above code we just written simple ("Chunks") & going one by one all of 
- this data, 8 sentences.

- But, here below code what we are trying to do is that, I am trying to "Enumerate". Enumerate basically does, 
- if we trying to pass a "List", so Enumeration will try to give a ID. 
- So lets suppose, we have List[] of the data inside a chunks, -> [ "for idx,chunk in enumerate(chunks): "] <-

- For every list, it will going to return the ("idx"), as well as it is going to return the ("Chunks").
- I am just trying to build my Meta information of the data. So that the region, that way I am trying to write.

- So here it is going to return the ("idx"), it is going to return the ("chunks") & this is the property of "Enumerate".
- This is the pattern -> indexes (0,1,2), this will going to return the dataset. 

# 1st Line code :

- Here, I have created 2 blank list[], keep it blanks.
- 1.Embedding List -> ( "emb_list = []" ) &
- 2.Meta -> Meta[]

# 1.Embedding List : 
- So, I have craeted Embedding_list[], where I will keep this all this embedding, which I am trying to produced.

- [ 0.01636502 -0.01330146  0.04360536 ... -0.00716132 -0.00025957 0.00777273]
- [-0.02828987 -0.00211549  0.03156904 ... -0.00953185 -0.01975844 -0.00468256]
- [-0.01842189 -0.03448532  0.01168446 ... -0.0130075   0.00758591 0.02325387] etc,.....

# 2.Meta List -> Meta[] :
- Anpther is Meta List[], where I am going to keep may be ( "IDs" ), maybe the "textual data", "Start Indexes" & "End Indexes" all of 
- such information. May be I can try t store over there. 

# 2nd line code : 
# Code : 
- for idx,chunk in enumerate(chunks):
-     vec = generate_embeddings(chunk)

- Here I will try to call a "generate Embedding" function ( ) & I am going to pass my "chunks", means I am going to pass my orginal data.
- This one -> "chunk" 1st line code we can see & then it is going to retun me vector. "Vector" means a Numerical Value. 

- This one -> [ 0.01636502 -0.01330146  0.04360536 ... -0.00716132 -0.00025957 0.00777273]
- [-0.02828987 -0.00211549  0.03156904 ... -0.00953185 -0.01975844 -0.00468256] etc,.....
- It is going to return, 

# 3rd & 4th line code :
- Code : vec = generate_embeddings(chunk)
-    emb_list.append(vec)

- So, here we have a vector, so here I am going to call embedding list.append what? append basically a Vector. SO, whatever it is trying 
- to generate try to append it.

# 5th Line code: 

- Code : emb_list.append(vec.astype("float32"))

- Before convert it, lets try to property type, -> ( "vec.astype" ) & then so whatever we are trying to append it. Keep it inside the 
- ( "Float32" ) bit format. "Numpy" compatiable & along with Facebook AI similarity search compatiable. whenever we try to store it,
- try to store at vector( "float32" ) bits.

- So this is going to basically keep on appending a data. It is going to append into 
- this, particular lists -> [ "emb_list = [] | meta = []" ] <- each and every data it is going to append.

# 6th Line Code :

- Now, we have to create a [ "Meta Information" ]. SO, I am going to call "meta.append". 
- So, here I am trying to append -> ( "id" ) |  I am trying to append "idx" a Text original data, that we have.
- This I am trying to tread as a "Text" data. So, it is going to be Meta Information.
- Lets execute this.
- Once, we execute this, it is going to create a vector & it is going to create and Meta information.
- All this information will be avaiable inside the Embedding List, & it will be avaiable inside the Meta.

- This below is the my final dataset.

In [16]:
emb_list = []
meta = []
for idx,chunk in enumerate(chunks):
    vec = generate_embeddings(chunk)
    emb_list.append(vec.astype("float32"))

    meta.append({"id": idx, "text": chunk})

- Now, if I will come & check my "Embedding List". 
- As a outcome, we can see I have all my data, avaiable in "Float32" format. And 


In [17]:
emb_list

[array([ 1.6688565e-02, -1.3501738e-02,  4.3314822e-02, ...,
        -7.1801147e-03, -1.3807718e-05,  7.8434953e-03],
       shape=(1536,), dtype=float32),
 array([-0.02824657, -0.00211815,  0.03152717, ..., -0.00951514,
        -0.01976704, -0.00461856], shape=(1536,), dtype=float32),
 array([-0.0184284 , -0.03449934,  0.01166509, ..., -0.0130235 ,
         0.00761142,  0.02324393], shape=(1536,), dtype=float32),
 array([-0.00206819, -0.01932828,  0.04296086, ..., -0.033941  ,
         0.00859491,  0.02564766], shape=(1536,), dtype=float32),
 array([ 0.01036524, -0.00490744,  0.05058131, ..., -0.0313709 ,
        -0.00884518,  0.00210155], shape=(1536,), dtype=float32),
 array([-0.01087056, -0.02549004,  0.03719883, ..., -0.01714737,
        -0.00013056,  0.00037601], shape=(1536,), dtype=float32),
 array([ 0.00250368, -0.02838254,  0.04988758, ...,  0.00022234,
        -0.02115859,  0.00431127], shape=(1536,), dtype=float32),
 array([ 0.04722496,  0.00496422,  0.04577735, ..., -0.016

- Now, If I will go ahead and check my "Meta", for all of this data.
- As a outcome, we can see that, we have the 
- 1.( Id is 0 or id : 0 ) & this was the Text -> "Making an Impact\nHelping Millions of Students....
- 2.( Id is 1 or id : 1 ) we have this text-> "1.5 million students from 34+ countries, providing them...
- 3.( Id is 2 or id : 2 ) we have this text-> 'backgrounds. They saw iNeuron as a lifeline—an opportunity.....

- I am just trying to maintain a proper Meta data information about my dataset. 

# Note : 
- So this was the the actual data -> 'id': 0, 'text': "Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment
- & this was the -> 
- [array([ 0.01636502, -0.01330146,  0.04360536, ..., -0.00716132, -0.00025957,  0.00777273], shape=(1536,), dtype=float32),
- array([-0.02828987, -0.00211549,  0.03156905, ..., -0.00953185, -0.01975844, -0.00468256], shape=(1536,), dtype=float32)
- this is basically respective conversion of those data. 

In [18]:
meta

[{'id': 0,
  'text': "Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in toda"},
 {'id': 1,
  'text': "1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.\nMany of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 202"},
 {'id': 2,
  'text': 'backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron '},
 {'id': 3,
  'tex

# Next: 
- Now we have the Meta information, we have the data which is avaiable into a vector format. Now what we have to perform?

# Next :

- So, we have this dataset, if we go & check what is the datatype.
- As we can see that this dataset is in the [ "List" ] format.
- Because, I have created simple List and inside that list, I was trying to keep all these data. 

In [19]:
type(emb_list)

list

# Note :  "Numpy Vector" format

- let's try to convert this dataset into a "Numpy Vector" format. So to do that I can try to call ( "Numpy".Vstack ) inside this,
- I will try to pass this entire List or (emb_list). my variable is ("xb").
- As a outcome, we can see that this entire vector converted into 2D vector, data inside the data. So we had list inside we had a "Array"
- I am just trying to convert that into a "Numpy compatiable" format.
- Now, Shape = (8, 1536), dtype=float32) <- means Now we have data of shape(8) into 8 rows & 1536 columns. 
- The length of the data is 1536 so that it is just trying to show us.

In [20]:
xb = np.vstack(emb_list)

# Next :
- Lets try to store this data into a variable called the ("xb").

In [21]:
xb

array([[ 1.6688565e-02, -1.3501738e-02,  4.3314822e-02, ...,
        -7.1801147e-03, -1.3807718e-05,  7.8434953e-03],
       [-2.8246567e-02, -2.1181451e-03,  3.1527173e-02, ...,
        -9.5151449e-03, -1.9767037e-02, -4.6185642e-03],
       [-1.8428400e-02, -3.4499343e-02,  1.1665091e-02, ...,
        -1.3023502e-02,  7.6114177e-03,  2.3243934e-02],
       ...,
       [-1.0870559e-02, -2.5490044e-02,  3.7198830e-02, ...,
        -1.7147368e-02, -1.3056057e-04,  3.7600618e-04],
       [ 2.5036815e-03, -2.8382542e-02,  4.9887575e-02, ...,
         2.2234007e-04, -2.1158595e-02,  4.3112719e-03],
       [ 4.7224957e-02,  4.9642199e-03,  4.5777347e-02, ...,
        -1.6907027e-02, -7.6696172e-02,  4.5340331e-03]],
      shape=(8, 1536), dtype=float32)

# Next :

- Now we are able to send this data into the "faiss" Facebook AI Similarity search, so that we have to even called the Normilization.
- So, fiass. try to call "normalized_l2" & I can try to pass my dataset, which is ("xb"). 
- so it will be able to normalized it. So here we are just trying to perform a normalization operation. 

In [22]:
import faiss

In [23]:
faiss.normalize_L2(xb)

# Next :

- Here variable = d & (xb).shape of [1]. as a outcome we can see that, (1536). Now that is going to be the dimension of 
- my vector database.
- I am just trying to find out the dimension from the dataset. 
- For 1 dataset dimension is basically 1536 & I need this dimension, to define my total index dimension
- inside the "faiss" Facebook AI Similarity search.

In [24]:
d = xb.shape[1]

In [25]:
d

1536

# Note : 

- Here what we can call is, there is something is called ["IndexFlatIP"] & inside that, we are going to pass a dimension ("d").
- so, Basically this one -> ("faiss") <- is going to set a dimension for a particular data, 
- because inside this "faiss" Facebook AI Similarity search, I can store a data, any of my dimension, maybe 2,000+, 3,000+, 5,000+.
- But the initially we have to define the dimension, ok fine what is the dimension of our dataset.
- our data can be using different different models, & every models give us different different dimension.

- 1.Code No.1 : 

- So, here I am going to set a dimension, & I am going to call it as a ( "Index" ) over here.
- Just set a dimension, Here instead of (d), I can write the (1536), which is my dimension and model is going to return, we can even
- do this one. this is completely fine. Now once we are able to create the "indexes".

- 2.Code No.2 :

- index.add() <- means index dot(.) add all the data at a time. Now once we are going to call the ( index.add ) & I will just going 
- to call my entire dataset this one -> (xb), which is again "Numpy Array" or "Numpy format" or Numpy compaitble data. It will be able to
- store all of these data automatically. 
- This below code, this is just one single line where we are trying to add all the data. After this step my dataset has been added.
- Now my dataset is store inside of "faiss" Facebook AI Similarity search.
- So, step wise somebody ask, then this 2 line of code is enough or required to store a data. 
- To store the data, just create the indexes & call add, then done.
- Its avaiable inside a memory, its avaiable inside the RAM, as of now. I can even try to create a Physical file and can try to store,
- this indexes, & I can try to store the "Meta Information" again which is going to discussed now only.

- Above codes we have discussed so far, we was just doing a "Data Preprocessing", preprocessing of the data.

- when I am trying to call "index.add", as of now everything in Memory only, it is not presisting my data into a hard disk into any 
- kind of physical file by the way, everything is avaiable into in Memory only. 
- To store any kind of data, only these 2 line of code is enough.
- Create the "Indexes", whatever dimension that we have that we can get from the model, & then call add work is done.

In [31]:
index = faiss.IndexFlatIP(1536)
index.add(xb)

# Note :

- As we discussed this is in Memory only, Let's suppose I would like to persist this entire indexes, in a file system. & apart from
- that, I would like to persist my "Meta Information" because even this "Meta information" is aviable in Variable.
- this one -> 
- {'id': 0, 'text': "Making an Impact\nHelping Millions ......},
- {'id': 1, 'text': "1.5 million students from 34+ countries,..."},
- {'id': 2,'text': 'backgrounds..... '},

- This Meta data also technically avaiable in memory only. Once I will restart my computer, so else I going to close this "VS code".
- After that I have to reexecute it, reload the entire things.
 

- Let's go and try to store this entire information, into some file.

# Next : 

- Code No.1 : 

- So, here for "Index", I am going to create some "variable" -> [ "index_path" ].
- Just try to store the "index of Vijay data Faiss". This is a file ["index_vijay.faiss" ] which I am going to create.
- this is the name of the file -> ("index_vijay") & this is the -> ("faiss") <- file extension.

- Now over here I am going to store all the indexes, all these vectors by the ways.

- Code No.2 :

- And "meta" information. Let's say variable is -> "meta_path" & I am going to create ("meta_vijay.jsonl") I am going to create &
- I am going to keep all those "meta data" -> 

- means this one ->
- 'id': 0, 'text': "Making an Impact\nHelping Millions.........
- 'id': 1, 'text': "1.5 million students from 34+ countries...........
- 'id': 2, 'text': 'backgrounds. They saw iNeuron.............
- 'id': 3,'text': "gnificant milestone, Sudhanshu remained focused on his mission........... etc......

- I can even attached date & time, if I want. I can even attached Geographical location, because with some data we can do that.
- we can attached longitude and latitude. case by case it depend. What kind of problem we can going to solve.

# Next : 

- So, I have created 2 files, as of now file path. 

- 1.Code No.1 : index_path = "index_vijay.faiss" <- Here we are going to store the indexes &
- 2.Code NO.2 : meta_path = "meta_vijay.json1" <- Here, I am going to store the Meta information.

In [32]:
index_path = "index_vijay.faiss"
meta_path = "meta_vijay.json1"

# Note : 
- Code No.1 :

- Here, If we have to store ("indexes"), SO simply I can call "Faiss". write indexes, 
- This is my index vaiable -> ["index"] & this is my Index path -> ["index_path"], if we execute this, we can see our folder section,
- new file has been created, name as -> [ "index_vijay.faiss" ]  a physical file, but we will not be able to read it out.
- we can see our folder section new file name has been created but, if we click on this file -> ("index_vijay.faiss") & try to open it
- we can't able to read that file, it will show my file but, we can't read it clearly, its a serialized file we will be able to find out.
- Technically all this data is avaiable over here, into a bytecode, it is a serialized one.
- ["index_vijay.faiss"] <- when we try to read the file, it tried to perform the deserialization.
- So, here we are able to create a physical file.
- so inside this physical file -> ["index_vijay.faiss"] <- I have all of these indxes, 

- This one check this line code -> [ emb_list ]
- array([ 1.6688565e-02, -1.3501738e-02,  4.3314822e-02, ...,-7.1801147e-03, -1.3807718e-05,  7.8434953e-03]),
- array([-0.02824689, -0.00209906,  0.03149973, ..., -0.0094944 ,-0.01982287, -0.00463947], shape=(1536,), dtype=float32),
- array([-0.01842189, -0.03448532,  0.01168446, ..., -0.0130075 , 0.00758591,  0.02325387], shape=(1536,), dtype=float32),
- array([-0.00206151, -0.01927499,  0.04288206, ..., -0.03394372, 0.0085956 ,  0.02564972], shape=(1536,), dtype=float32)

- Now I am able to process my data  into hard disk, even if I am going to close my computer that is completely fine.
- If I will open this file, -> ["index_vijay.faiss"] , I dnt have to go through all these lengthy process, which I have already done.
- I can directly start from this particular file -> ["index_vijay.faiss"] & searching for the data. In a physical file, I can store
- the data in a as a indexes.

In [33]:
faiss.write_index(index, index_path)

# Note : Meta information

- Let's suppose, If I have store "meta indexes" information as well. I can write the code and store the information.Similar to above one
- Lets try to import JSON, & operating system, I am going to import.


In [34]:
import json,os

- Lets support, I am trying to store the "Meta" information, into a Physical file. 

- I can try to write ( with open ) meta_pathn "w" or write.
- SO, we have this ["meta"]

- this one -> "meta"
- {'id': 0, 'text': "Making an Impact\nHelping Millions ......},
- {'id': 1, 'text': "1.5 million students from 34+ countries,..."},
- {'id': 2,'text': 'backgrounds..... '},

- And, I am going to write this inside this file, now if I will execute this below code then, as a outcome, we see on our folder section,
- it created a file name as -> ["meta_vijay.json1"] & if we double click and open it, we can see we can see the we have the same
- "Meta" information, inside this file -> ["meta_vijay.json1"].
- So, the "meta" Information I am able to store, Even my embedding, I am able to store it.

In [35]:
with open(meta_path, "w") as f:
    for item in meta:
        f.write(json.dumps(item) + "\n")

# Note : creating physical file and able to store inside my physical hard-Disk

- So, we are able to store both "meta" & "vector"

- 1."index_vijay.faiss"
- 2."meta_vijay.json1"

- Now, we can see in the left side, folder or file section, these 2 has been created.
- Now this is avaiable into my physical hark Disk. I am able to create a physical file.

- Now, even after close my system and come back, I will be able to get the exact same thing.
- So, storing is done. We are able to store our dataset.

# Note :  Search Operation :

- Let's say I am going to write a something is like, ["who is Vijay"]? obviously from that information from that embedding.
- it must give me some answer, it must give me some return. Now lets do the Search opeartion.
- So, that I can write the query and it will give me in return some result.

# For Exampple : 
- 1.query = "what was the mission of iNeuron?"
- 2.what has happened in 2022.
- 3.May be who is "Sudhanshu kumar?

- So, we all know this information is already avaiable in that particular dataset.

- Lets say: 
- Somebody write query, my query is this one -> ["query = "what was the mission of iNeuron?"] <- 1st of all I will convert this query.
- into a embeddings. So, I am going to follow the exact same thing done even before.

- So, here I am going to generating Embedding for this particular query -> ["query = "what was the mission of iNeuron?"].

- Code No.2 : ["generate_embeddings(query)"]

- once I am able to generate the embedding -> ["generate_embeddings(query)"] &

- Code No.3 : [ "astype("float32")"]

- I wil convert this into a ("float 32"), because I have done that even before.

- Code No.4 : [ reshape(1, -1) ]

- Then I can try to call a reshape, so that then my dataset will be flat.

- So, this will be going to my actual query -> ["generate_embeddings(query)"] & suppose I will name it as ("q")
- This is my query and I am not able to convert into it into like a some Normalization format. 

- code No.5 : faiss.normalize_L2(q)

- Now just try to call, ( "faiss.normalize_L2(q)" ) <- and try to pass the query, so that it will be able to calculate the norm.
- In a denominator, we have the normalization. we can say this is equivalent distance. It will be able to calculate norm over here.

- Code N0.6 :

- Then I can try to call the "index.search(q, 5), search wise I am trying to pass my query ("q") & I am trying to pass top K values.
- means try to give me top ( 5 ) result or I can say give me top 3 result. That is also completely.
- So, this is ( 5 ) basically a top "K".

- whichever will be having, top 3 most like maxmimum similarity, So it is going to give me that result.

# Next :
- If we execute this code, as a outcome, we can see it is able to give me some short result.
- & ( "Index.search" ) wise, I am able to see some outcome, I am able to see some result.
- index.search(q, 3) then, I will get Array No.[ 2,1,0 ] or I can try index.search(q, 5), then I will get something like -
- Array No.2, Array No.1, & Array No.(0), Array No.3 & Array No.4 . This is the result which it is trying to give it to me.
- Here is a score -> (array([[0.5985576 , 0.5481801 , 0.48705295, 0.3973113 , 0.20701069], in a decending order, it is trying to give me
- the score. & in particular score, it is trying to give it to me an Arrays -> ([[2, 1, 0, 3, 4]].

# Below, what does this mean actually :

- The 2nd score index is the best, this is what it actually means because similarity score is basically 0.5985576 or (0.59).
- 1.So, Array No.[2] is the best, & 
- 2.then 2nd best Array is No.[1]. & 
- 3.then 3rd best Array is No.[0] &
- 4.then 4th best Array is No.[3] &
- 5.then 5th best Array is No.[4]

- This is what it is trying to give it to me.
- So, here we are able to get the result ->  array [2, 1, 0, 3, 4] & now we can do more in details coding and we can able to see 
- a result which will be Human readable.
- As of now it is trying to give it to me "Cosign" similarity score & after that it is trying to give me respective indexes.
- For this particular Array indexes, this is the similary score, which they are able to find out & 
- this is the arrangement in a decending order.

In [36]:
query = "what was the mission of iNeuron?"
q = generate_embeddings(query).astype("float32").reshape(1, -1)
faiss.normalize_L2(q)
index.search(q, 5)

(array([[0.59863365, 0.548123  , 0.48909983, 0.3973113 , 0.20664331]],
       dtype=float32),
 array([[2, 1, 0, 3, 4]]))

# Continue... 1:27:05 PM

- Lets suppose, if we have to arrange this entire data, -> array[2, 1, 0, 3, 4] <- & we have to showcase, what is avaiable at
- Array index No.[2] ?
- Just left corner click on this file name -> ["meta_vijay.json1"] <- inside this we are able to see the Meta information 
- -> {"id": 2, "text": "backgrounds. They saw iNeuron as a lifeline\u2014an opportunity to rise above their circums...<- we can go & check
- Even I can write python code, which I will be able to print, Becasue I have ID's even over here inside this -> ["meta_vijay.json1"] <-
- So, Array No.[2], means "ID:2" {"id": 2, "text": "backgrounds. They saw iNeuron as a lifeline\u2014an opportunity ...... &
- Array No.[4], means "ID:4" {"id": 4, "text": "reneur and Teacher: Sudhanshu's Dual Legacy\nSudhanshu's journey......
- If we do a python coding and get to know, which line is mathching is with my line.

# Note :
- what was my input -> [ query = "what was the mission of iNeuron?" ]
- So, If I will go and check inside -> ["meta_vijay.json1"] <- we can see inside this 
- Line No.-> {"id": 2, "text": "backgrounds. They saw iNeuron as a lifeline\u2014an opportunity......} <- it is giving me accurate
- result, although match score is [0.5985576 ] or (0.59),  which is closer to 1, but yes this is the -> [iNeuron?], most similar
- search, I am able to get.

# Next : Let's say I will asked ["Who is Sudhanshu"?]

- As a outcome, we can see I am able to find top score is Array No.[6] index by the way.
- If we open this file left corner -> ["meta_vijay.json1"] <- as we can see top score is 
- this line No.-> {"id": 6, "text": "accessible to learners at all levels.\n\nHis commitment to....} is matching, which is mention.
- We are able to get the desire result.

- Now, I am able to store the information, after scoring the information. I am able to even "Query" the information. Its look like
- that I am able to correct result.

In [29]:
query = "who is sudhanshu?"
q = generate_embeddings(query).astype("float32").reshape(1, -1)
faiss.normalize_L2(q)
index.search(q, 5)

(array([[0.5228137 , 0.5069984 , 0.50195414, 0.48918533, 0.35711852]],
       dtype=float32),
 array([[6, 5, 4, 3, 1]]))

# Public Question :

- Faiss normalized L2 query embedding, please explain why we applied?
- Check my Notebook, I have written over there.

# Normalization ?

- Let's discussed with respect to a "Numpy" Array.
- Let's suppose I consider Numpy Array over here, inside this Array lets me take small, data lets say [4,5,6,7,8] this is something
- inside "Numpy" Array.
- I will keep inside variable name is -> ["Test"] <-
- Now, I will convert this dataset into the "float32"
- Now If I am going to execute, and I am going to call my "Test" data. As a outcome we can see that,
- this is how my "Test" dataset looks like. -> [array([[4., 5., 6., 7., 8.]], dtype=float32)] <- after the data, 
- there will be a floating point number.

In [37]:
test = np.array([[4,5,6,7,8]], dtype=np.float32)

In [38]:
test

array([[4., 5., 6., 7., 8.]], dtype=float32)

# Here, If I am going to call a normalization !

- Let's suppose -> np.linalg.norms(test) <- So it is giving me -> (13.784049) or 13.78 something like this, it is return to me.

- See this is the differnece, I have taken one random array -> array([[4,5,6,7,8]] with respect to a vector.
- If we try to find out the norms, means, if we are going to find out -> ||X|| (√4²,5²,6²,7²,8²) <- this is what it is going to written.
- so this is something it is trying to return -> [ np.float32(13.784049) ]. & this is a Norm -> ||X||.

In [39]:
np.linalg.norm(test)

np.float32(13.784049)

- Now, If I will going to call -> faiss.normalize_L2 & then If I am going to pass my "test" data.
- What it will going to return, here it will try to pass. -> np.linalg.norm(test) <- As a outcome we can see it is closer to [1].

- Now, when I am going and try to apply a normalization with L2 & L2 is nothing but a regulation technique.
- if we check above one -> here lenth is (13.784049) or 13 % with same data.
- But, when I call the L2 on top of this this one -> "normalize" <- it has normalized each and every dataset to a unit value.
- only [1].
- So, that when I try to find out, a "Cosine Similarity". As we know the formula : 
- cosine = (a.b / ||a|| x ||b||). So what we are trying to say is that, lets normalizing, lets make it [1 & 1 ] this one -> ||a|| x ||b||
- Technically, we are nullifying the entire denominators.SO our Cosin similarity score we will be able to find out, just by finding out,
- dot(.) product & this is what this IP is doing.
- So "faiss index flatIP" that we have called it from the "Faiss" databased.
- Technically, we have called the libary called it "index flatIP". Technically it is trying to calculate the "indexflatIP".
- IP or (inner product).
- So, it is trying to calculate the "Index Flat IP". So, it just trying to calculate a (axb), & it do not support to calculate the
- demoninator one. which is ||a|| x ||b||.
- Region is simple, We have already Normalized it, means our denominator is always [1], for any vector.
- Any vector we will going to paas, after normalization, if we are going to calculate the "Norm".
- SO, after ( L2 ), if we are going to calculate the Norms, it will be always [1].
- So, ||a|| x ||b|| or [ 1x 1 ] is going to be [1] only. 
- So, that is the region, we are try to calling a normalization over here. This is how this "Index.FlatIP" is going to work.
- Technically, it will calculates just inner product IP or ( a x b ), it will going to calculate. 
- we are just conveting this denominator into [1], so that it is not going to effect it & ("index flatIP") bydefault,
- it is going to calculate dot(.) of it.
- This is how this libary has been build.

In [40]:
faiss.normalize_L2(test)

In [41]:
np.linalg.norm(test)

np.float32(0.99999994)

# Python Mapping

- Now we can do a "Python Mapping", with respect to this "Meta Information" & then we will be able to get the Final result.
- This is the one of the DataBase. but we have alot of others DataBase as well.

# Note : DataBases :

- Go to Google and search for these DataBases :

- 2.Qdrant DataBase URL : https://qdrant.tech/
- 3..Pinecone DataBase  URL : https://www.pinecone.io/
- 4.chromaDB DataBase URL : https://www.trychroma.com/
- 5.weaviate DataBase URL : https://weaviate.io/

# Note :

- So, we have so many different different Databases, which is avaiable in a same segment, Some of these database going to provide you.
- even an Manage Hosting, they have already hosted over the cloud platform, we don't have do anything, we have just go and do the 
- installation and start storing our dataset.